### PR TITLE
Resiliency suite for PDS and first Test Case

### PR DESCRIPTION
**Pull request by dbhatnagar-px**
Checking in Resiliency Suite along with a test case of rebooting a node during deployment. 
1. This Test case is basic framework for Resiliency like how we want to achieve resiliency execution in context of PDS
2. It validates if this is a resiliency test case
3. Based out of "n" different type of supported failures, this test case picks one failure (as of now it's just one)
4. It checks if that particular failure condition is met
5. Once failure condition is met, this test case induces that particular failure
6. It checks for any errors and fails the TC if there is any failure that happens

Have run this again the test scenario where we want to reboot a node onto which a deployment pod is coming up. Following are the exact steps this test case has: 
1. Create a Data Service
2. Check if data service is coming up 
3. Select a node onto which a pod is yet to come up
4. Reboot that node
5. After rebooting, validate the data service still comes up fine 
6. Validate configuration of data service

Jenkins Run: https://jenkins.pwx.dev.purestorage.com/job/PDS/job/pds-branch-build-test/147/

### CODE REVIEW

File: tests/pds/dataservice_test.go
Comment (Madanagopal19) [inline tests/pds/dataservice_test.go:356]: Is this change Intentional ?

**Review by Madanagopal19 — CHANGES_REQUESTED** [review]

File: tests/pds/dataservice_test.go
Comment (dbhatnagar-px) [reply tests/pds/dataservice_test.go]: Removed.

**Review by gauraa2 — CHANGES_REQUESTED** [review]

File: tests/pds/dataservice_test.go
Comment (Madanagopal19) [inline tests/pds/dataservice_test.go:320]: Incase of error in this method, are we not failing it immediately?

File: tests/pds/dataservice_test.go
Comment (Madanagopal19) [inline tests/pds/dataservice_test.go:323]: I can see `DefineFailureType ` is updating the var `FailureType ` and `FailureType ` is not consumed by any other places. Is this expected ?

**Review by Madanagopal19 — CHANGES_REQUESTED** [review]

File: drivers/pds/lib/utils.go
Comment (Madanagopal19) [inline drivers/pds/lib/utils.go:775]: lets have log.InfoD

File: drivers/pds/lib/resiliency_utils.go
Comment (Madanagopal19) [reply drivers/pds/lib/resiliency_utils.go:43]: Lets also have a new name instead of `DefineFailureType`

**Review by Madanagopal19 — CHANGES_REQUESTED** [review]

File: drivers/pds/lib/utils.go
Comment (dbhatnagar-px) [reply drivers/pds/lib/utils.go:752]: Not Reqd now with channel changes

File: drivers/pds/lib/utils.go
Comment (dbhatnagar-px) [reply drivers/pds/lib/utils.go]: This comment is not reqd now

File: tests/pds/dataservice_test.go
Comment (dbhatnagar-px) [reply tests/pds/dataservice_test.go]: As discussed, it is failing immediately. Call to function and error catching is happening later than this line

File: tests/pds/dataservice_test.go
Comment (dbhatnagar-px) [reply tests/pds/dataservice_test.go]: Failuretype was the struct that was used to call it's failure method. I have updated the names now as we discussed

**Comment by dbhatnagar-px** [comment]
Have addressed review comments except one - As of now it's only running for postgres for minimal testing. Once PR is ready to merge, I will simply remove that check and merge it in. 
Latest Jenkins run after addressing all review comments and also design change to incorporate channels instead of flags : 
https://jenkins.pwx.dev.purestorage.com/job/PDS/job/pds-branch-build-test/154/

File: tests/pds/dataservice_test.go
Comment (dbhatnagar-px) [reply tests/pds/dataservice_test.go]: Keeping this comment as unresolved for now - this is for minimal testing. Will merge the PR after removing this check only.

**Review by Madanagopal19 — APPROVED** [review]
LGTM

File: tests/pds/resiliency_tests.go
Comment (Madanagopal19) [inline tests/pds/resiliency_tests.go:39]: as discussed pass the replicas as params

**Review by Madanagopal19 — CHANGES_REQUESTED** [review]
Minor changes

**Review by gauraa2 — APPROVED** [review]
LGTM

File: tests/pds/resiliency_tests.go
Comment (dbhatnagar-px) [reply tests/pds/resiliency_tests.go]: Done

**Review by gauraa2 — APPROVED** [review]
LGTM, please fix the build issue.

File: drivers/pds/lib/resiliency_utils.go
Comment (lsrinivas-pure) [inline drivers/pds/lib/resiliency_utils.go:20]: please use camelcase format

File: drivers/pds/lib/resiliency_utils.go
Comment (lsrinivas-pure) [inline drivers/pds/lib/resiliency_utils.go:30]: please use camcelcase format

File: drivers/pds/lib/utils.go
Comment (lsrinivas-pure) [inline drivers/pds/lib/utils.go:173]: I think 5 seconds is long time for replicas to be up, Is 1 second ok?

File: tests/pds/resiliency_tests.go
Comment (lsrinivas-pure) [inline tests/pds/resiliency_tests.go:21]: any reason why this test is limited to postgres?

File: drivers/pds/lib/resiliency_utils.go
Comment (lsrinivas-pure) [inline drivers/pds/lib/resiliency_utils.go:79]: can you add some checks here to close the channel gracefully without causing panic

File: drivers/pds/lib/resiliency_utils.go
Comment (lsrinivas-pure) [inline drivers/pds/lib/resiliency_utils.go:105]: are you validating this error  any where?

File: drivers/pds/lib/utils.go
Comment (lsrinivas-pure) [inline drivers/pds/lib/utils.go:799]: return this err and handle it

File: drivers/pds/lib/resiliency_utils.go
Comment (dbhatnagar-px) [reply drivers/pds/lib/resiliency_utils.go:114]: Yes, this is getting validated when all functions execute in parallel and proper error is being returned to the test case

File: tests/pds/resiliency_tests.go
Comment (dbhatnagar-px) [reply tests/pds/resiliency_tests.go:21]: This is intentional for testing purposes. I will remove this check while doing the final merge. It will run for all data services.

File: drivers/pds/lib/resiliency_utils.go
Comment (dbhatnagar-px) [reply drivers/pds/lib/resiliency_utils.go]: Done

File: drivers/pds/lib/resiliency_utils.go
Comment (dbhatnagar-px) [reply drivers/pds/lib/resiliency_utils.go]: Done

File: drivers/pds/lib/resiliency_utils.go
Comment (dbhatnagar-px) [reply drivers/pds/lib/resiliency_utils.go]: Done. Although it's not required as of now as we are calling close during Post Test Execution, but it's better to have checks. Thanks for pointing it out - we will keep this in mind during further test case addition

File: drivers/pds/lib/utils.go
Comment (dbhatnagar-px) [reply drivers/pds/lib/utils.go]: Done

File: tests/pds/dataservice_test.go
Comment (dbhatnagar-px) [reply tests/pds/dataservice_test.go]: Resolving this. Ready for final merge.

**Review by Madanagopal19 — APPROVED** [review]
lgtm

File: drivers/pds/lib/resiliency_utils.go
Comment (lsrinivas-pure) [inline drivers/pds/lib/resiliency_utils.go:70]: will this not return error?

File: drivers/pds/lib/resiliency_utils.go
Comment (lsrinivas-pure) [inline drivers/pds/lib/resiliency_utils.go:96]: how are you handling the error returned here?

File: drivers/pds/lib/resiliency_utils.go
Comment (lsrinivas-pure) [inline drivers/pds/lib/resiliency_utils.go:53]: if any function returns error, how are you handling it?

File: drivers/pds/lib/resiliency_utils.go
Comment (dbhatnagar-px) [reply drivers/pds/lib/resiliency_utils.go:70]: It will be a channel based return from next test (PR) onwards

File: drivers/pds/lib/resiliency_utils.go
Comment (dbhatnagar-px) [reply drivers/pds/lib/resiliency_utils.go:96]: It's getting handled as a separate error check condition once the go routine exits.

**Review by lsrinivas-pure — APPROVED** [review]
lgtm

**Pull request merged**
